File: socket.c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <dirent.h>
#include <netdb.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>
/*******
 * TCP *
 *******/
void socket_set_block(int socket,int on) {
    int flags;
    flags = fcntl(socket,F_GETFL,0);
    if (on==0) {
        fcntl(socket, F_SETFL, flags | O_NONBLOCK);
    }else{
        flags &= ~ O_NONBLOCK;
        fcntl(socket, F_SETFL, flags);
    }
}
int socket_connect(const char *host,int port,int timeout){
    struct sockaddr_in sa;
    struct hostent *hp;
    int sockfd = -1;
    hp = gethostbyname(host);
    if(hp==NULL){
        return -1;
    }
    bcopy((char *)hp->h_addr, (char *)&sa.sin_addr, hp->h_length);
    sa.sin_family = hp->h_addrtype;
    sa.sin_port = htons(port);
    sockfd = socket(hp->h_addrtype, SOCK_STREAM, 0);
    socket_set_block(sockfd,0);
    connect(sockfd, (struct sockaddr *)&sa, sizeof(sa));
    fd_set          fdwrite;
    struct timeval  tvSelect;
    FD_ZERO(&fdwrite);
    FD_SET(sockfd, &fdwrite);
    tvSelect.tv_sec = timeout;
    tvSelect.tv_usec = 0;
    int retval = select(sockfd + 1,NULL, &fdwrite, NULL, &tvSelect);
    if (retval<0) {
        close(sockfd);
        return -2;
    }else if(retval==0){//timeout
        close(sockfd);
        return -3;
    }else{
        int error=0;
        int errlen=sizeof(error);
        getsockopt(sockfd, SOL_SOCKET, SO_ERROR, &error, (socklen_t *)&errlen);
        if(error!=0){
            close(sockfd);
            return -4;//connect fail
        }
        socket_set_block(sockfd, 1);
        int set = 1;
        setsockopt(sockfd, SOL_SOCKET, SO_NOSIGPIPE, (void *)&set, sizeof(int));
        return sockfd;
    }
}
int socket_close(int socketfd){
    return close(socketfd);
}
int socket_read(int socketfd,char *data,int len){
    int readlen=(int)read(socketfd,data,len);
    return readlen;
}
int socket_send(int socketfd,const char *data,int len){
    int byteswrite=0;
    while (len-byteswrite>0) {
        int writelen=(int)write(socketfd, data+byteswrite, len-byteswrite);
        if (writelen<0) {
            return -1;
        }
        byteswrite+=writelen;
    }
    return byteswrite;
}
//return socket fd  listen函数在一般在调用bind之后-调用accept之前调用
int socket_listen(const char *addr,int port){
    //create socket
    int socketfd=socket(AF_INET, SOCK_STREAM, 0);
    int reuseon   = 1;
    setsockopt( socketfd, SOL_SOCKET, SO_REUSEADDR, &reuseon, sizeof(reuseon) );
    //bind
    struct sockaddr_in serv_addr;
    memset( &serv_addr, '\0', sizeof(serv_addr));
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_addr.s_addr = inet_addr(addr);
    serv_addr.sin_port = htons(port);
    int r=bind(socketfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr));
    if(r==0){
        if (listen(socketfd, 128)==0) {
            return socketfd;
        }else{
            return -2;//listen error
        }
    }else{
        return -1;//bind error
    }
}
//return client socket fd
int socket_accept(int onsocketfd,char *remoteip,int* remoteport){
    socklen_t clilen;
    struct sockaddr_in  cli_addr;
    clilen = sizeof(cli_addr);
    int newsockfd = accept(onsocketfd, (struct sockaddr *) &cli_addr, &clilen);
    char *clientip=inet_ntoa(cli_addr.sin_addr);
    memcpy(remoteip, clientip, strlen(clientip));
    *remoteport=cli_addr.sin_port;
    if(newsockfd>0){
        return newsockfd;
    }else{
        return -1;
    }
}



/*******
 * TCP *
 *******/


int datagramsocket(const char *addr,int port){
    //create socket
    int socketfd = socket(AF_INET,SOCK_DGRAM,0);
    int reuseon = 1;
    setsockopt( socketfd, SOL_SOCKET, SO_REUSEADDR, &reuseon, sizeof(reuseon) );
    //bind
    if (port != 0){
        struct sockaddr_in serv_addr;
        memset( &serv_addr, '\0', sizeof(serv_addr));
        serv_addr.sin_family = AF_INET;
        serv_addr.sin_addr.s_addr = inet_addr(addr);
        serv_addr.sin_port = htons(port);
        if (0 == bind(socketfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr))) {
            return socketfd;
        }else{
            return -1;
        }
    }else{
        return socketfd;
    }
}



int datagramsocket_recive(int socket_fd,char *outdata,int expted_len,char *remoteip,int* remoteport){
    struct sockaddr_in  cli_addr;
    socklen_t clilen=sizeof(cli_addr);
    memset(&cli_addr, 0x0, sizeof(struct sockaddr_in));
    int len=(int)recvfrom(socket_fd, outdata, expted_len, 0, (struct sockaddr *)&cli_addr, &clilen);
    char *clientip=inet_ntoa(cli_addr.sin_addr);
    memcpy(remoteip, clientip, strlen(clientip));
    *remoteport=cli_addr.sin_port;
    return len;
}



int datagramsocket_get_server_ip(char *host,char *ip){
    struct hostent *hp;
    struct sockaddr_in addr;
    hp = gethostbyname(host);
    if(hp==NULL){
        return -1;
    }
    bcopy((char *)hp->h_addr, (char *)&addr.sin_addr, hp->h_length);
    char *clientip=inet_ntoa(addr.sin_addr);
    memcpy(ip, clientip, strlen(clientip));
    return 0;
}
//send message to addr and port
int datagramsocket_send(int socket_fd,char *msg,int len, char *toaddr, int topotr){
    struct sockaddr_in addr;
    socklen_t addrlen=sizeof(addr);
    memset(&addr, 0x0, sizeof(struct sockaddr_in));
    addr.sin_family=AF_INET;
    addr.sin_port=htons(topotr);
    addr.sin_addr.s_addr=inet_addr(toaddr);
    int sendlen=(int)sendto(socket_fd, msg, len, 0, (struct sockaddr *)&addr, addrlen);
    return sendlen;
}

File: Socket.swift

import Foundation

@asmname("socket_connect") func c_socket_connect(host:UnsafePointer<Int8>,port:Int32,timeout:Int32) -> Int32
@asmname("socket_close") func c_socket_close(fd:Int32) -> Int32
@asmname("socket_send") func c_socket_send(fd:Int32,buff:UnsafePointer<UInt8>,len:Int32) -> Int32
@asmname("socket_read") func c_socket_read(fd:Int32,buff:UnsafePointer<UInt8>,len:Int32) -> Int32
@asmname("socket_listen") func c_socket_listen(addr:UnsafePointer<Int8>,port:Int32)->Int32
@asmname("socket_accept") func c_socket_accept(onsocketfd:Int32,ip:UnsafePointer<Int8>,port:UnsafePointer<Int32>) -> Int32

public class connect{
    var addr:String
    var port:Int
    var fd:Int32?
    init(){
        self.addr = "0.0.0.0"
        self.port = 0
    }
    init(port: Int){
        self.addr = "0.0.0.0"
        self.port = port
    }
    init(addr: String,port: Int){
        self.addr = addr
        self.port = port
    }
}

public class Socket:connect{
    override init() {
        super.init()
    }
    //禁用
    private override init(port: Int) {
        super.init(port: port)
    }
    override init(addr: String, port: Int) {
        super.init(addr: addr, port: port)
        var result = connect(15)
        println(result.1)
    }
    private func connect(timeout: Int)->(Bool,String){
        var result_fd = c_socket_connect(self.addr,Int32(self.port),Int32(timeout))
        if result_fd > 0{
            self.fd = result_fd
            return (true,"connect success")
        }else{
            switch result_fd{
            case -1:
                return (false,"qeury server fail")
            case -2:
                return (false,"connection closed")
            case -3:
                return (false,"connect timeout")
            default:
                return (false,"unknow err.")
            }
        }
    }
    /*
    * close socket
    * return success or fail with message
    */
    public func close()->(Bool,String){
        if let fd = self.fd{
            c_socket_close(fd)
            self.fd = nil
            return (true,"close success")
        }else{
            return (false,"socket not open")
        }
    }
    /*
    * send data
    * return success or fail with message
    */
    public func send(data: [UInt8])->(Bool,String){
        if let fd = self.fd{
            var sendsize = c_socket_send(fd,data, Int32(data.count))
            if Int(sendsize) == data.count{
                return (true,"send success")
            }else{
                return (false,"send error")
            }
        }else{
            return (false,"socket not open")
        }
    }
    /*
    * send string
    * return success or fail with message
    */
    public func send(str: String)->(Bool,String){
        if let fd = self.fd{
            var sendsize = c_socket_send(fd, str, Int32(strlen(str)))
            if sendsize==Int32(strlen(str)){
                return (true,"send success")
            }else{
                return (false,"send error")
            }
        }else{
            return (false,"socket not open")
        }
    }
    /*
    *
    * send nsdata
    */
    public func send(data: NSData)->(Bool,String){
        if let fd = self.fd{
            var buff:[UInt8] = [UInt8](count:data.length,repeatedValue:0)
            data.getBytes(&buff, length: data.length)
            var sendsize = c_socket_send(fd, buff, Int32(data.length))
            if sendsize == Int32(data.length){
                return (true,"send success")
            }else{
                return (false,"send error")
            }
        }else{
            return (false,"socket not open")
        }
    }
    /*
    * read data with expect length
    * return success or fail with message
    */
    public func read(expectlen:Int)->[UInt8]?{
        if let fd = self.fd{
            var buff:[UInt8] = [UInt8](count:expectlen,repeatedValue:0)
            var readLen = c_socket_read(fd, &buff, Int32(expectlen))
            if readLen<=0{
                return nil
            }
            var rs = buff[0...Int(readLen-1)]
            var data:[UInt8] = Array(rs)
            return data
        }
       return nil
    }
}

public class SocketServer:connect{
    override init(port: Int) {
        super.init(port: port)
        listen()
    }
    override init(addr: String, port: Int) {
        super.init(addr: addr, port: port)
        listen()
    }

    private func listen()->(Bool,String){
        var fd = c_socket_listen(self.addr, Int32(self.port))
        if fd>0{
            self.fd = fd
            return (true,"listen success")
        }else{
            return (false,"listen fail")
        }
    }
    public func accept()->Socket?{
        if let serverfd = self.fd{
            var buff:[Int8] = [Int8](count:16,repeatedValue:0)
            var port:Int32 = 0
            var clientfd = c_socket_accept(serverfd, &buff,&port)
            if clientfd<0{
                return nil
            }
            var socket: Socket = Socket()
            socket.fd = clientfd
            socket.port = Int(port)
            if let addr = String(CString: buff, encoding: NSUTF8StringEncoding){
               socket.addr = addr
            }
            return socket
        }
        return nil
    }
    public func close()->(Bool,String){
        if let fd = self.fd{
            c_socket_close(fd)
            self.fd=nil
            return (true,"close success")
        }else{
            return (false,"socket not open")
        }
    }
}


